File: app/src/main/java/com/stedi/randomimagegenerator/app/model/data/generatorparams/base/GeneratorParams.java
package com.stedi.randomimagegenerator.app.model.data.generatorparams.base;

import android.os.Parcelable;
import android.support.annotation.NonNull;

import com.stedi.randomimagegenerator.app.model.data.GeneratorType;
import com.stedi.randomimagegenerator.app.model.data.generatorparams.ColoredCirclesParams;
import com.stedi.randomimagegenerator.app.model.data.generatorparams.ColoredNoiseParams;
import com.stedi.randomimagegenerator.app.model.data.generatorparams.ColoredPixelsParams;
import com.stedi.randomimagegenerator.app.model.data.generatorparams.ColoredRectangleParams;
import com.stedi.randomimagegenerator.app.model.data.generatorparams.FlatColorParams;
import com.stedi.randomimagegenerator.app.model.data.generatorparams.MirroredParams;
import com.stedi.randomimagegenerator.app.model.data.generatorparams.TextOverlayParams;
import com.stedi.randomimagegenerator.generators.Generator;

public abstract class GeneratorParams implements Parcelable {
    @NonNull
    public Generator getGenerator() {
        return GeneratorProxyKt.proxy(createGenerator());
    }

    public abstract void setId(int id);

    public abstract int getId();

    public abstract boolean isEditable();

    @NonNull
    public abstract GeneratorType getType();

    @NonNull
    public static GeneratorParams createDefaultParams(@NonNull GeneratorType type) {
        if (type.isEffect())
            throw new IllegalArgumentException("type must not be effect");

        switch (type) {
            case FLAT_COLOR:
                return new FlatColorParams();
            case COLORED_PIXELS:
                ColoredPixelsParams pixelsParams = new ColoredPixelsParams();
                pixelsParams.setValue(10);
                return pixelsParams;
            case COLORED_CIRCLES:
                ColoredCirclesParams circlesParams = new ColoredCirclesParams();
                circlesParams.setValue(50);
                return circlesParams;
            case COLORED_RECTANGLE:
                ColoredRectangleParams rectangleParams = new ColoredRectangleParams();
                rectangleParams.setValue(50);
                return rectangleParams;
            case COLORED_NOISE:
                return new ColoredNoiseParams();
            default:
                throw new IllegalStateException("unreachable code");
        }
    }

    @NonNull
    public static EffectGeneratorParams createDefaultEffectParams(@NonNull GeneratorType effectType, @NonNull GeneratorParams target) {
        if (!effectType.isEffect())
            throw new IllegalArgumentException("type must be effect");

        switch (effectType) {
            case MIRRORED:
                return new MirroredParams(target);
            case TEXT_OVERLAY:
                return new TextOverlayParams(target);
            default:
                throw new IllegalStateException("unreachable code");
        }
    }

    @NonNull
    protected abstract Generator createGenerator();

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        return true;
    }

    @Override
    public String toString() {
        return "GeneratorParams{" +
                "getType()=" + getType() +
                ", isEditable()=" + isEditable() +
                '}';
    }

    @Override
    public int describeContents() {
        return 0;
    }
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/model/data/generatorparams/base/GeneratorParams.kt
package com.stedi.randomimagegenerator.app.model.data.generatorparams.base

import android.os.Parcel
import android.os.Parcelable
import android.support.annotation.CallSuper
import com.j256.ormlite.field.DatabaseField
import com.stedi.randomimagegenerator.Rig
import com.stedi.randomimagegenerator.app.model.data.GeneratorType
import com.stedi.randomimagegenerator.app.model.data.generatorparams.*
import com.stedi.randomimagegenerator.generators.Generator

@Suppress("EqualsOrHashCode")
abstract class GeneratorParams : Parcelable {

    @DatabaseField(generatedId = true, columnName = "id")
    var id: Int = 0

    // OrmLite required
    constructor()

    companion object {
        const val COLORED_PIXELS_DEFAULT_MULTIPLIER = 10
        const val COLORED_NOISE_DEFAULT_MULTIPLIER = 4

        fun createRandomDefaultParams(): GeneratorParams {
            return createDefaultParams(GeneratorType.NON_EFFECT_TYPES.let { it[Math.round(Rig.random((it.size - 1).toFloat()))] })
        }

        fun createDefaultParams(type: GeneratorType): GeneratorParams {
            if (type.isEffect) {
                throw IllegalArgumentException("type must not be effect")
            }

            when (type) {
                GeneratorType.FLAT_COLOR -> return FlatColorParams()
                GeneratorType.COLORED_PIXELS -> {
                    val pixelsParams = ColoredPixelsParams()
                    pixelsParams.setValue(COLORED_PIXELS_DEFAULT_MULTIPLIER)
                    return pixelsParams
                }
                GeneratorType.COLORED_CIRCLES -> {
                    return ColoredCirclesParams()
                }
                GeneratorType.COLORED_LINES -> {
                    return ColoredLinesParams()
                }
                GeneratorType.COLORED_RECTANGLE -> {
                    return ColoredRectangleParams()
                }
                GeneratorType.COLORED_NOISE -> {
                    val noiseParams = ColoredNoiseParams()
                    noiseParams.setValue(COLORED_NOISE_DEFAULT_MULTIPLIER)
                    return noiseParams
                }
                else -> throw IllegalStateException("unreachable code")
            }
        }

        fun createDefaultEffectParams(effectType: GeneratorType, target: GeneratorParams): EffectGeneratorParams {
            if (!effectType.isEffect) {
                throw IllegalArgumentException("type must be effect")
            }

            when (effectType) {
                GeneratorType.MIRRORED -> return MirroredParams(target)
                GeneratorType.THRESHOLD -> return ThresholdParams(target)
                GeneratorType.TEXT_OVERLAY -> return TextOverlayParams(target)
                else -> throw IllegalStateException("unreachable code")
            }
        }
    }

    fun getGenerator() = proxy(createGenerator())

    abstract fun getType(): GeneratorType

    protected abstract fun createGenerator(): Generator

    @CallSuper
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        return !(other == null || javaClass != other.javaClass)
    }

    override fun toString() = "GeneratorParams{" +
            "getType()=${getType()}" +
            '}'

    @CallSuper
    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(this.id)
    }

    protected constructor(parcel: Parcel) {
        this.id = parcel.readInt()
    }

    override fun describeContents() = 0
}